### PR TITLE
Add a carousel of bonus levels from previous stages

### PR DESCRIPTION
**Pull request by balderdash**
The buttons scroll through the stages, which include all the stages with bonus levels up to and including the current stage.

screenshot from 2017-11-10 10-39-49

Clicking 'try it' will take you to a level page that has the stage header for the stage you came from (not the stage the level was actually in), and finishing the puzzle will bring you back to the lesson extras page you were on.

Mark had a couple more suggestions that will come in another PR:
* The green/white checkmark overlay should be a smaller, but still obvious, badge
* Do something to the 'try it' button for completed levels to reinforce the message that the level was previously completed

### CODE REVIEW

File: apps/src/code-studio/components/stageExtras/BonusLevels.jsx [hunk 1]
```diff
@@ -1,19 +1,23 @@
+import FontAwesome from '../../../templates/FontAwesome';
```
Comment (joshlory) [inline apps/src/code-studio/components/stageExtras/BonusLevels.jsx]: Tiny nit: can also do `@cdo/apps/templates/FontAwesome` here (will preserve reference even if this file is moved).

File: apps/src/code-studio/components/stageExtras/BonusLevels.jsx [hunk 2]
```diff
@@ -24,14 +28,38 @@ const styles = {
   challengeRow: {
     clear: 'both',
     overflow: 'hidden',
+    display: 'inline-block',
+    position: 'relative',
+    whiteSpace: 'normal',
+    transition: 'left 0.25s ease-out',
```
Comment (joshlory) [inline apps/src/code-studio/components/stageExtras/BonusLevels.jsx:34]: 👍

File: apps/src/code-studio/reporting.js [hunk 3]
```diff
@@ -290,6 +290,12 @@ reporting.sendReport = function (report) {
           var fallback = getFallbackResponse(report) || {};
           response.redirect = fallback.redirect;
         }
+        if (appOptions.isBonusLevel) {
+          // Bonus levels might have to take students back to a different stage,
+          // ignore the redirect in response and use the url from appOptions
+          // instead
+          response.redirect = appOptions.nextLevelUrl;
```
Comment (joshlory) [inline apps/src/code-studio/reporting.js:297]: Where does `nextLevelUrl` come from?

File: dashboard/app/models/script.rb [hunk 4]
```diff
@@ -422,6 +422,20 @@ def get_script_level_by_chapter(chapter)
     script_levels[chapter - 1] # order is by chapter
   end
 
+  def get_bonus_script_levels(current_stage)
+    unless @all_bonus_script_levels
+      @all_bonus_script_levels = stages.map do |stage|
+        {
+          stageNumber: stage.absolute_position,
```
Comment (joshlory) [inline dashboard/app/models/script.rb]: Relative or absolute position here?

File: apps/src/code-studio/reporting.js (hunk 3 above)
Comment (balderdash) [reply apps/src/code-studio/reporting.js:297]: https://github.com/code-dot-org/code-dot-org/pull/19033/files#diff-1f68e16ab551449bdbd0fb7f01004f68R144

It's just always passed in appOptions now.

File: dashboard/app/models/script.rb (hunk 4 above)
Comment (balderdash) [reply dashboard/app/models/script.rb]: Whoops, yeah that should be relative.

**Pull request merged**